pkg/plugins/gateway/algorithms/prefix_cache.go

```diff
@@ -18,113 +18,205 @@ package routingalgorithms
+
+		if len(matchedPods) > 0 {
+			targetPod = getTargetPodFromMatchedPods(p.cache, readyPods, matchedPods)
+			if targetPod != nil {
+				klog.InfoS("prefix_cache_matched_pods",
+					"request_id", ctx.RequestID,
+					"target_pod", targetPod.Name,
+					"target_pod_ip", targetPod.Status.PodIP,
+					"matched_pods", matchedPods,
+					"pod_request_count", getRequestCounts(p.cache, readyPods))
+			} else {
+				klog.InfoS("prefix_cache_skip_matched_pods",
+					"request_id", ctx.RequestID,
+					"matched_pods", matchedPods,
+					"pod_request_count", getRequestCounts(p.cache, readyPods))
+			}
+		}
 	}
 
-	var matchedPodNames, readyPodNames []string
-	for _, p := range matchedPods {
-		matchedPodNames = append(matchedPodNames, p.Status.PodIP)
+	// no pod with prefix match, as a fallback select pod with least request count
+	if len(matchedPods) == 0 || targetPod == nil {
+		targetPod = selectTargetPodWithLeastRequestCount(p.cache, readyPods)
+		klog.InfoS("prefix_cache_fallback_least_request_count",
+			"request_id", ctx.RequestID,
+			"target_pod", targetPod.Name,
+			"target_pod_ip", targetPod.Status.PodIP,
+			"matched_pods", matchedPods,
+			"pod_request_count", getRequestCounts(p.cache, readyPods))
 	}
-	for _, p := range readyPods {
-		readyPodNames = append(readyPodNames, p.Status.PodIP)
+
+	if len(prefixHashes) > 0 {
+		p.prefixCacheIndexer.AddPrefix(prefixHashes, ctx.Model, targetPod.Name)
 	}
-	klog.InfoS("prefix cache route",
-		"matched_tokens", matchedTokens,
-		"unmatched_tokens", unMatchedTokens,
-		"matched_pods", matchedPodNames,
-		"ready_pods", readyPodNames,
-		"target_pod", targetPod.Status.PodIP)
 
 	ctx.SetTargetPod(targetPod)
 	return ctx.TargetAddress(), nil
 }
+
+func getTargetPodFromMatchedPods(cache cache.Cache, readyPods []*v1.Pod, matchedPods map[string]int) *v1.Pod {
+	var targetPodName string
+	requestCount := []float64{}
+
+	podRequestCount := getRequestCounts(cache, readyPods)
+	for _, cnt := range podRequestCount {
+		requestCount = append(requestCount, float64(cnt))
+	}
+	meanRequestCount := mean(requestCount)
+	stdDevRequestCount := standardDeviation(requestCount)
+
+	podnames := []string{}
+	for podname := range matchedPods {
+		podnames = append(podnames, podname)
+	}
+	rand.Shuffle(len(podnames), func(i, j int) {
+		podnames[i], podnames[j] = podnames[j], podnames[i]
+	})
+
+	// sort pods with decreasing %perfixmatch AND for same %prefixmatch sort by increasing request count
+	sort.SliceStable(podnames, func(i, j int) bool {
+		if matchedPods[podnames[i]] == matchedPods[podnames[j]] {
+			return podRequestCount[podnames[i]] < podRequestCount[podnames[j]]
+		}
+		return matchedPods[podnames[i]] > matchedPods[podnames[j]]
+	})
+
+	// select targetpod with highest %prefixmatch and request_count within stddev
+	for _, podname := range podnames {
+		reqCnt := float64(podRequestCount[podname])
+		if reqCnt <= meanRequestCount+stdDevRequestCount {
```

Comment:
can we run some unit test to approve this is efficiency? do we need some factor on `std` etc?

Reply:
If I am understanding correctly, you question is why one std_dev and it can potentially be 2 or 3 standard deviations as well?

Unit test can not prove the efficiency, because higher the standard deviation is more requests can be packed on same pod with prefix_match. 

One alternative is to have add configurable multiply factor to standard deviation and default value can be 1.

Reply:
Added a factor for std deviation. For benchmark test, a factor of 2 improves the TTFT by 10% compare to factor of 1, update the PR description with details.